Unable to open up jar file when going to definition (for dependencyContents)

Arch Linux,
neovim 0.5.0

```
❯ clojure-lsp --version       
clojure-lsp 2021.07.19-14.46.52
clj-kondo 2021.06.18
```

Hi,

On trying to executed `clojure/dependencyContents` on, for example `clojure.string/join`, I receive these two errors (note that the exceptions, one after the other, are subtly different):

```
2021-07-28T07:05:52.642Z daffy ERROR [clojure-lsp.server:?] - 
com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine      PosixJavaThreads.java:  192
            com.oracle.svm.core.thread.JavaThreads.threadStartRoutine           JavaThreads.java:  553
                        java.util.concurrent.ForkJoinWorkerThread.run  ForkJoinWorkerThread.java:  183
                          java.util.concurrent.ForkJoinPool.runWorker          ForkJoinPool.java: 1594
                               java.util.concurrent.ForkJoinPool.scan          ForkJoinPool.java: 1656
             java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec          ForkJoinPool.java: 1020
                             java.util.concurrent.ForkJoinTask.doExec          ForkJoinTask.java:  290
              java.util.concurrent.CompletableFuture$AsyncSupply.exec     CompletableFuture.java: 1692
               java.util.concurrent.CompletableFuture$AsyncSupply.run     CompletableFuture.java: 1700
                  clojure-lsp.server.LSPTextDocumentService/reify/get                 server.clj:  192
                                       clojure-lsp.handlers/code-lens               handlers.clj:  263
                    clojure-lsp.feature.code-lens/reference-code-lens              code_lens.clj:   41
                                     clojure-lsp.shared/uri->filename                 shared.clj:  100
                                                  java.net.URL.<init>                   URL.java:  488
                                                  java.net.URL.<init>                   URL.java:  541
                                                  java.net.URL.<init>                   URL.java:  674
                            sun.net.www.protocol.jar.Handler.parseURL               Handler.java:  163
                   sun.net.www.protocol.jar.Handler.parseAbsoluteSpec               Handler.java:  181
java.lang.NullPointerException: no !/ in spec
java.net.MalformedURLException: no !/ in spec

2021-07-28T07:10:56.144Z daffy DEBUG [clojure-lsp.server:?] - :codeLens 1ms
2021-07-28T07:10:56.144Z daffy DEBUG [clojure-lsp.server:?] - :documentHighlight 1ms
2021-07-28T07:10:57.643Z daffy ERROR [clojure-lsp.server:?] - 
com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine      PosixJavaThreads.java:  192
            com.oracle.svm.core.thread.JavaThreads.threadStartRoutine           JavaThreads.java:  553
                        java.util.concurrent.ForkJoinWorkerThread.run  ForkJoinWorkerThread.java:  183
                          java.util.concurrent.ForkJoinPool.runWorker          ForkJoinPool.java: 1594
                               java.util.concurrent.ForkJoinPool.scan          ForkJoinPool.java: 1656
             java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec          ForkJoinPool.java: 1020
                             java.util.concurrent.ForkJoinTask.doExec          ForkJoinTask.java:  290
              java.util.concurrent.CompletableFuture$AsyncSupply.exec     CompletableFuture.java: 1692
               java.util.concurrent.CompletableFuture$AsyncSupply.run     CompletableFuture.java: 1700
                  clojure-lsp.server.LSPTextDocumentService/reify/get                 server.clj:  212
                            clojure-lsp.handlers/semantic-tokens-full               handlers.clj:  272
                      clojure-lsp.feature.semantic-tokens/full-tokens        semantic_tokens.clj:  110
                                     clojure-lsp.shared/uri->filename                 shared.clj:  100
                                                  java.net.URL.<init>                   URL.java:  488
                                                  java.net.URL.<init>                   URL.java:  541
                                                  java.net.URL.<init>                   URL.java:  674
                            sun.net.www.protocol.jar.Handler.parseURL               Handler.java:  163
                   sun.net.www.protocol.jar.Handler.parseAbsoluteSpec               Handler.java:  181
java.lang.NullPointerException: no !/ in spec
java.net.MalformedURLException: no !/ in spec
```

The URI that is being passed to clojure-lsp is like this:

`
jar:file:///home/david/.m2/repository/org/clojure/clojure/1.10.3/clojure-1.10.3.jar!/clojure/core.clj
`

Which is invoked via this function:

```
function! LoadClojureContent(uri)
    setfiletype clojure
    let content = CocRequest('clojure-lsp', 'clojure/dependencyContents', {'uri': a:uri})
    call setline(1, split(content, "\n"))
    setl nomodified readonly
endfunction
```

More Details.

Since the change to `coc-clojure` as the language server (rather than using `clojure-lsp` directly, I've discovered I need to change the function LoadClojureContent (detailed above) to have this:

`CocRequest('clojure', ........` and not `CocRequest('clojure-lsp', ......`)

Doing that gets a step further, but then I observe a stack trace:

```
2021-07-28T10:31:22.025 ERROR (pid:148481) [attach] - Request error: sendRequest [
  'clojure',
  'clojure/dependencyContents',
  {
    uri: 'jar:file:///home/david/.m2/repository/org/clojure/clojure/1.10.3/clojure-1.10.3.jar!/clojure/string.clj'
  }
] ag [Error]: Internal error.
    at $l (/home/david/.env/vim/bundle/coc.nvim/build/index.js:36:224)
    at Kh (/home/david/.env/vim/bundle/coc.nvim/build/index.js:35:11261)
    at Immediate.<anonymous> (/home/david/.env/vim/bundle/coc.nvim/build/index.js:35:11111)
    at processImmediate (node:internal/timers:464:21) {
  code: -32603,
  data: 'java.lang.RuntimeException: java.lang.reflect.InvocationTargetException\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)\n' +
    '\tat java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)\n' +
    '\tat java.util.concurrent.FutureTask.run(FutureTask.java:264)\n' +
    '\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n' +
    '\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n' +
    '\tat java.lang.Thread.run(Thread.java:829)\n' +
    '\tat com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:553)\n' +
    '\tat com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)\n' +
    'Caused by: java.lang.reflect.InvocationTargetException\n' +
    '\tat java.lang.reflect.Method.invoke(Method.java:566)\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)\n' +
    '\t... 13 more\n' +
    "Caused by: java.lang.IllegalStateException: Attempting to call unbound fn: #'clojure-lsp.main/extension\n" +
    '\tat clojure.lang.Var$Unbound.throwArity(Var.java:45)\n' +
    '\tat clojure.lang.AFn.invoke(AFn.java:36)\n' +
    '\tat clojure.lang.Var.invoke(Var.java:388)\n' +
    '\tat clojure_lsp.ClojureExtensions.dependencyContents(ClojureExtensions.java:26)\n' +
    '\t... 15 more\n'
```

--- a/resources/META-INF/native-image/clojure-lsp/clojure-lsp/reflect-config.json
+++ b/resources/META-INF/native-image/clojure-lsp/clojure-lsp/reflect-config.json
@@ -63,6 +63,10 @@
   "allPublicMethods":true
 },
 {
+  "name":"java.net.URLDecoder",
+  "allPublicMethods":true
+},
+{
   "name":"java.util.ArrayList",
   "allPublicConstructors":true
 },

--- a/src/clojure_lsp/shared.clj
+++ b/src/clojure_lsp/shared.clj
@@ -7,7 +7,8 @@
    [clojure.string :as string]
    [taoensso.timbre :as log])
   (:import
-   [java.net URI URL JarURLConnection]
+   [java.net URI URL JarURLConnection URLDecoder]
+   [java.nio.charset StandardCharsets]
    [java.nio.file Paths]))
 
 (defn assoc-some
@@ -92,13 +93,25 @@
   (-> uri Paths/get .toString
       (string/replace #"^[a-z]:\\" string/upper-case)))
 
+(defn- unescape-uri
+  [uri]
+  (try
+    (URLDecoder/decode uri (.name StandardCharsets/UTF_8)) ;; compatible with Java 1.8 too!
+    (catch UnsupportedOperationException e
+      (log/warn "Unable to decode URI. Returning URI as-is." e)
+      uri)
+    (catch IllegalArgumentException e
+      (log/warn "Unable to decode URI. Returning URI as-is." e)
+      uri)))
+
 (defn uri->filename
   "Converts a URI string into an absolute file path.
 
   The output path representation matches that of the operating system."
   [^String uri]
   (if (string/starts-with? uri "jar:")
-    (let [conn ^JarURLConnection (.openConnection (URL. uri))
+    (let [unescaped-uri (unescape-uri uri)
+          conn ^JarURLConnection (.openConnection (URL. unescaped-uri))
           jar-file (uri-obj->filepath ^URI (.toURI ^URL (.getJarFileURL conn)))]
       (str jar-file ":" (.getEntryName conn)))
     (let [uri-obj (URI. uri)

--- a/test/clojure_lsp/shared_test.clj
+++ b/test/clojure_lsp/shared_test.clj
@@ -83,3 +83,16 @@
             :end   {:line      0
                     :character 0}}
            (shared/->range {:row 0 :end-row 0 :col 0 :end-col 0})))))
+
+(def unescape-uri #'shared/unescape-uri)
+
+(deftest unescape-uri-test
+  (testing "URI should unescape."
+    (is (= "jar:file:///home/foo/bar.jar!baz.clj"
+           (unescape-uri "jar:file%3A///home/foo/bar.jar%21baz.clj"))))
+  (testing "URI should remain the same."
+    (is (= "file:///home/foo/bar.jar"
+           (unescape-uri "file:///home/foo/bar.jar"))))
+  (testing "URI should remain the same as IllegalArgumentException is thrown."
+    (is (= "file:///home/foo/bar.jar%%"
+           (unescape-uri "file:///home/foo/bar.jar%%")))))

